File: java/other/BigInteger.java
package other;

import java.util.Arrays;

public class BigInteger {
    public String add(String num1, String num2) {
        StringBuffer result = new StringBuffer();
        int carry = 0;
        for (int i = 1; i <= Math.max(num1.length(), num2.length()); ++i) {
            int value1 = (i <= num1.length()) ? num1.charAt(num1.length() - i) - '0' : 0;
            int value2 = (i <= num2.length()) ? num2.charAt(num2.length() - i) - '0' : 0;
            int value = value1 + value2 + carry;
            carry = value / 10;
            result.append(value % 10);
        }
        if (0 != carry) {
            result.append(carry);
        }
        return result.reverse().toString();
    }

    public String subtract(String num1, String num2) {
        if (!greater(num1, num2)) {
            return "-" + subtract(num2, num1);
        }

        StringBuffer result = new StringBuffer();
        int borrow = 0;
        for (int i = 1; i <= num1.length(); ++i) {
            int value1 = num1.charAt(num1.length() - i) - '0';
            int value2 = (i <= num2.length()) ? num2.charAt(num2.length() - i) - '0' : 0;
            int value = value1 - value2 - borrow;
            borrow = 0;
            if (value < 0) {
                value += 10;
                borrow = 1;
            }
            result.append(value);
        }

        int index = result.length() - 1;
        while (index > 0 && '0' == result.charAt(index)) {
            --index;
        }
        return result.reverse().toString().substring(result.length() - 1 - index);
    }

    public String multiply(String num1, String num2) {
        char[] values = new char[num1.length() + num2.length()];
        Arrays.fill(values, '0');
        for (int i = 1; i <= num1.length(); ++i) {
            int value1 = num1.charAt(num1.length() - i) - '0';
            for (int j = 1; j <= num2.length(); ++j) {
                int value2 = num2.charAt(num2.length() - j) - '0';
                int value = value1 * value2;
                int index = i + j - 2;
                addValues(values, index, value % 10);
                addValues(values, index + 1, value / 10);
            }
        }

        int index = values.length - 1;
        while (index > 0 && '0' == values[index]) {
            --index;
        }
        StringBuffer result = new StringBuffer();
        for (int i = index; i > 0; --i) {
            result.append(values[i]);
        }
        return result.toString();
    }

    public String[] divide(String num1, String num2) {
        String[] result = new String[2];
        Arrays.fill(result, "0");
        while (greater(num1, num2)) {
            int diff = num1.length() - num2.length();
            String aux1 = fillTail(num2, diff);
            String aux2 = fillTail("1", diff);
            if (!greater(num1, aux1)) {
                aux1 = aux1.substring(0, aux1.length() - 1);
                aux2 = aux2.substring(0, aux2.length() - 1);
            }
            while (greater(num1, aux1)) {
                num1 = subtract(num1, aux1);
                result[0] = add(result[0], aux2);
            }
        }
        result[1] = num1;
        return result;
    }

    private void addValues(char[] values, int index, int value) {
        values[index] += value;
        if (values[index] > '9') {
            values[index] -= 10;
            ++values[index + 1];
        }
    }

    private boolean greater(String num1, String num2) {
        if (num1.length() != num2.length()) {
            return num1.length() > num2.length();
        }
        return num1.compareTo(num2) >= 0;
    }

    private String fillTail(String num, int length) {
        return (0 != length) ? num + String.format("%0" + length + "d", 0) : num;
    }
}

File: py/other/BigInteger.py
#!/usr/bin/env python
# -*- coding:gbk -*-

from typing import List

class BigInteger:
    def add(self, num1: str, num2: str) -> str:
        result, carry = '', 0
        for i in range(1, max(len(num1), len(num2)) + 1):
            value1, value2 = ord(num1[-1 * i]) - ord('0') if i <= len(num1) else 0, ord(num2[-1 * i]) - ord('0') if i <= len(num2) else 0
            value = value1 + value2 + carry
            carry = value // 10
            result += str(value % 10)
        if carry:
            result += str(carry)
        return result[ : : -1]

    def subtract(self, num1: str, num2: str) -> str:
        if not self.__greater(num1, num2):
            return '-' + self.subtract(num2, num1)

        result, borrow = '', 0
        for i in range(1, len(num1) + 1):
            value1, value2 = ord(num1[-1 * i]) - ord('0'), ord(num2[-1 * i]) - ord('0') if i <= len(num2) else 0
            value = value1 - value2 - borrow
            borrow = 0
            if value < 0:
                value += 10
                borrow = 1
            result += str(value)

        index = len(result) - 1
        while index > 0 and '0' == result[index]:
            index -= 1
        return result[index : : -1]

    def multiply(self, num1: str, num2: str) -> str:
        values = [0 for _ in range(len(num1) + len(num2))]
        for i, value1 in enumerate(num1[ : : -1]):
            for j, value2 in enumerate(num2[ : : -1]):
                value, index = (ord(value1) - ord('0')) * (ord(value2) - ord('0')), i + j
                self.__addValues(values, index, value % 10)
                self.__addValues(values, index + 1, value // 10)

        index = len(values) - 1
        while index > 0 and 0 == values[index]:
            index -= 1
        result = ''
        for value in values[index : : -1]:
            result += str(value)
        return result

    def divide(self, num1: str, num2: str) -> List[str]:
        result = ['0' for _ in range(2)]
        while self.__greater(num1, num2):
            diff = len(num1) - len(num2)
            aux1, aux2 = self.__fillTail(num2, diff), self.__fillTail('1', diff)
            if not self.__greater(num1, aux1):
                aux1, aux2 = aux1[ : -1], aux2[ : -1]
            while self.__greater(num1, aux1):
                num1, result[0] = self.subtract(num1, aux1), self.add(result[0], aux2)
        result[1] = num1
        return result

    def __addValues(self, values: List[int], index: int, value: int) -> None:
        values[index] += value
        if values[index] > 9:
            values[index] -= 10
            values[index + 1] += 1

    def __greater(self, num1: str, num2: str) -> bool:
        if len(num1) != len(num2):
            return len(num1) > len(num2)
        return num1 >= num2

    def __fillTail(self, num: str, length: int) -> str:
        return num + ('0' * length)
